### DISCUSSION TITLE
Tile image decode using web workers

### DISCUSSION THREAD

**Discussion by 2mer** [openseadragon/openseadragon#2554] [Show and tell]
I looked into other image viewers to see how they tackle the decode of images (cornerstone, OHIF viewer - which uses cornerstone under the hood)

they seem to decode in web workers powered by WASM

I didnt get to benchmark this too much, but here is an example repository (that should be taken with a grain of salt) highlighting a simple (and totally naive) implementation for patching a DZI tilesource (the example only works with JPEGs and no WASM) to get its tile data via a web worker instead of directly requesting the tiles

[demo](https://2mer.github.io/osd-worker-example/)
[source](https://github.com/2mer/osd-worker-example/tree/main)

Im not 100% comfortable with this hacky/patchy solution because it does feel like it could probably be done using the "correct" data flow, and not by replacing some members on the tile source after-the-fact, but surprisingly due to the fact it was implemented in this patchy approach it led to a pretty cool interaction - the patch can be performed at multiple levels, allowing fine grained control for opting into worker decode:
- patch a single tile source instance - instance level
- patch all tile sources added to a viewer - viewer level
- patch `OpenSeadragon.DziTileSource.prototype` (or any other prototype) - implementation level

**Comment by msalsbery** [comment]
Very cool!

Not too helpful for image decoding using what the browser already natively supports (like your JPEG example) but definitely a benefit doing non-native stuff. 

Unsupported image types for tiles could be decoded pretty efficiently with web workers and WASM...just nobody has asked me for that yet 😜

**Comment by iangilman** [comment]
Awesome... Thank you for sharing this here! This could be quite useful for someone who needs this.
